From 8c59c9a699ee84cbc13a558f1093eb8f69ce99ac Mon Sep 17 00:00:00 2001
From: Yan Zhao <yan.y.zhao@intel.com>
Date: Fri, 18 Sep 2020 20:41:31 +0800
Subject: [PATCH] qemu auto launch script for WMI device

add --hp-wmi-mediation option for WMI device

Change-Id: I92216bdf9a79cd12c736ab9f55d4c6df33e9fe60
Tracked-On: OAM-92883
Signed-off-by: Yan Zhao <yan.y.zhao@intel.com>
Signed-off-by: Swee Yee Fonn <swee.yee.fonn@intel.com>

diff --git a/scripts/setup_host.sh b/scripts/setup_host.sh
index b60181b..8d77773 100755
--- a/scripts/setup_host.sh
+++ b/scripts/setup_host.sh
@@ -28,7 +28,7 @@ function ubu_changes_require(){
     if [ x$res = xn ]; then
         return 1
     fi
-    sudo apt install -y wget mtools ovmf dmidecode python3-usb python3-pyudev pulseaudio jq acpica-tools
+    sudo apt install -y wget mtools ovmf dmidecode python3-usb python3-pyudev pulseaudio jq acpica-tools acpid

     # Install libs for vm-manager
     sudo apt install -y libglib2.0-dev libncurses-dev libuuid1 uuid-dev libjson-c-dev
diff --git a/scripts/start_civ.sh b/scripts/start_civ.sh
old mode 100755
new mode 100644
index 7259fbe..30a16ea
--- a/scripts/start_civ.sh
+++ b/scripts/start_civ.sh
@@ -25,6 +25,7 @@ GUEST_RPMB_DEV_PID=
 GUEST_RPMB_DEV_SOCK=
 GUEST_THERMAL_DAEMON_PID=
 GUEST_BATTERY_DAEMON_PID=
+GUEST_WMI_LISTENER_DAEMON_PID=
 GUEST_IMAGE=$WORK_DIR/android.qcow2
 GUEST_VSOCK="-device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid=3"
 GUEST_SHARE_FOLDER=
@@ -53,6 +54,7 @@ GUEST_AAF_DIR=$WORK_DIR/aaf
 GUEST_AAF_CONFIG=$GUEST_AAF_DIR/mixins.spec
 GUEST_POWER_BUTTON=
 GUEST_GRAPHIC_MODE=
+GUEST_WMI_DEV=

 GUEST_STATIC_OPTION="\
  -name caas-vm \
@@ -153,6 +155,28 @@ function setup_swtpm() {
     swtpm socket --tpmstate dir=$WORK_DIR/vtpm0 --tpm2 --ctrl type=unixio,path=$WORK_DIR/vtpm0/swtpm-sock &
 }

+function setup_hp_wmi_listener() {
+    #start listening daemon for wmi
+    local wmi_listener_daemon=$SCRIPTS_DIR/wmi-listen.sh
+    local guest_wmi_qmp_sock=$WORK_DIR/qmp-wmi-sock
+    local guest_wmi_target="95F24279-4D7B-"
+    if [ -f $wmi_listener_daemon ]; then
+        modprobe acpi-mediator
+        setsid $SCRIPTS_DIR/wmi-listen.sh "$guest_wmi_qmp_sock" "$guest_wmi_target" "$SCRIPTS_DIR" &> $WORK_DIR/wmi-listener.log &
+        GUEST_WMI_LISTENER_DAEMON_PID=$!
+        #echo wmid $GUEST_WMI_LISTENER_DAEMON_PID
+        GUEST_WMI_DEV="-qmp unix:$guest_wmi_qmp_sock,server,nowait -device acpi-proxy -smbios \"type=11,value=FBYTE#6S6b7B7H7M7R7W7m7s8YaBapaqaub8bTbhcAdRdUdqfAjAjQjq.M2;,value=BUILDID#18WWCABN6AN#SABA#DABA;,value=EDK2_1,value=Buff=2,value=Dock Embedded Controller FW Version:21.53;,value=Dock Serial Number:5CG8396B3Y;,value=HRDWFEATS=VTX:1;VTD:1;SGX:0;NONHPBATDET:1\""
+    fi
+}
+
+function cleanup_hp_wmi_listener() {
+    local procname=`basename "$0"`
+    kill_daemon_proc "$GUEST_WMI_LISTENER_DAEMON_PID" "$procname"
+    if [[ ! -z "$GUEST_WMI_LISTENER_DAEMON_PID" ]]; then
+        pkill -TERM -P "$GUEST_WMI_LISTENER_DAEMON_PID"
+    fi
+}
+
 function setup_thermal_mediation() {
     local thermal_daemon=$SCRIPTS_DIR/thermsys
     if [ -f $thermal_daemon ]; then
@@ -683,6 +707,7 @@ function cleanup() {
     cleanup_rpmb_dev
     cleanup_thermal_mediation
     cleanup_battery_mediation
+    cleanup_hp_wmi_listener
     [[ -z $GUEST_USB_PT_DEV ]] || set_pt_usb unset
     [[ -z $GUEST_AUDIO_PT_DEV ]] || set_pt_audio unset
     [[ -z $GUEST_UDC_PT_DEV ]] || set_pt_udc unset
@@ -739,6 +764,7 @@ function launch_guest() {
               $GUEST_KIRQ_CHIP \
               $GUEST_STATIC_OPTION \
               $GUEST_EXTRA_QCMD \
+              $GUEST_WMI_DEV \
     "

     echo $EXE_CMD
@@ -913,6 +939,10 @@ function parse_arg() {
                 set_guest_pwr_vol_button
                 ;;

+            --hp-wmi-mediation)
+                setup_hp_wmi_listener
+                ;;
+
             -?*)
                 echo "Error: Invalid option $1"
                 show_help
--
2.29.2

